### PR TITLE
Display a warning if the endpoint uses the HTTP protocol

### PR DESCRIPTION
**Pull request by axbarsan**
References https://github.com/giantswarm/giantswarm/issues/8502

### CODE REVIEW

**Comment by axbarsan** [comment]
**A potential problem**

On some `gsctl` commands (the ones which have pre-run validation, such as `info` ), the warning is shown twice.
The reason for this is that `collectArguments()`, the place where the endpoint selection and normalization is done, gets run twice (once in the pre-run validation, and once in the actual command).

**Preview** 


Screenshot 2020-01-24 at 11 19 12

**Comment by oponder** [comment]
Nice! thanks @axbarsan 

Is there maybe another place we can put it to avoid the double error?

Alternatively, how many places have a pre-validation, maybe we can adjust something in those cases to avoid the double error? Like making sure the second validation call gets the adjusted https endpoint and so no warning appears on the second time through?

**Comment by axbarsan** [comment]
> Nice! thanks @axbarsan
> 
> Is there maybe another place we can put it to avoid the double error?
> 
> Alternatively, how many places have a pre-validation, maybe we can adjust something in those cases to avoid the double error? Like making sure the second validation call gets the adjusted https endpoint and so no warning appears on the second time through?

There are currently 19 commands that use pre-run validation. As for exact placement in the codebase, on the current design, placing it in the`normalizeEndpoint` function makes the most sense to me.

A potential solution is storing the arguments in a package-level variable, so they wouldn't have to be computed twice, and the warning would be only shown once

**Comment by marians** [comment]
I don't mind the double warning. If that's the easiest right now. We are hopefully dealing with an edge case.

Regarding automatic conversion: please keep in mind that this library is not only made for gsctl (that's why it's been extracted from gsctl into a separate lib). Is there a use case to consider where connecting to HTTP instead of HTTPS would make sense? I hope no. I'm fine with trying and finding out after merge.

**Comment by axbarsan** [comment]
> I don't mind the double warning. If that's the easiest right now. We are hopefully dealing with an edge case.
> 
> Regarding automatic conversion: please keep in mind that this library is not only made for gsctl (that's why it's been extracted from gsctl into a separate lib). Is there a use case to consider where connecting to HTTP instead of HTTPS would make sense? I hope no. I'm fine with trying and finding out after merge.

I can't think of a use case that would require HTTP.
But what do you think about storing the arguments at package level for the commands that use pre-run validation?
Would mean handling the storage of the arguments in the pre-validation function, and then in the actual command function, we'd just get the arguments from the variable. Tried it now and it works well.

Although that is not directly related to this PR, as that is gsctl territory

**Comment by marians** [comment]
Package level variable sounds good to me. Can you create and link the according gsctl PR, too?

**Comment by axbarsan** [comment]
Removed auto-adjusting the protocol because it would cause errors in gsctl tests, as those use HTTP by default, and enablind HTTPS would be too big of a change for such a small thing

**Comment by axbarsan** [comment]
Here are the changes to gsctl
https://github.com/giantswarm/gsctl/pull/499

File: config/config.go [hunk 1]
```diff
@@ -701,11 +702,17 @@ func normalizeEndpoint(u string) string {
...
+	isHttp := strings.HasPrefix(u, "http://")
+
 	// if URL has no scheme, prefix it with the default scheme.
-	if !strings.HasPrefix(u, "https://") && !strings.HasPrefix(u, "http://") {
+	if !strings.HasPrefix(u, "https://") && !isHttp {
 		u = "https://" + u
 	}
 
+	if isHttp {
+		fmt.Println(color.YellowString("[Warning] Endpoint URL uses an insecure protocol"))
```
Comment (marians) [inline config/config.go]:
```suggestion
		fmt.Println(color.YellowString("Warning: endpoint URL uses an insecure protocol"))
```

Little style alignment.

**Review by marians — APPROVED** [review]

**Pull request merged**